exporter/prometheus.go

```diff
@@ -0,0 +1,455 @@
+			Namespace: namespace,
+			Name:      "network_transmit_packets_dropped_total",
+			Help:      "Cumulative count of bytes transmitd",
+		},
+			[]string{"name", "id", "image"},
+		),
+		networkTxErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "network_transmit_errors_total",
+			Help:      "Cumulative count of errors encountered",
+		},
+			[]string{"name", "id", "image"},
+		),
+
+		tasksSleeping: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "tasks_sleeping",
+			Help:      "Number of sleeping tasks",
+		},
+			[]string{"name", "id", "image"},
+		),
+		tasksRunning: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "tasks_running",
+			Help:      "Number of running tasks",
+		},
+			[]string{"name", "id", "image"},
+		),
+		tasksStopped: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "tasks_stopped",
+			Help:      "Number of stopped tasks",
+		},
+			[]string{"name", "id", "image"},
+		),
+		tasksUninterruptible: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "tasks_uinterruptible",
+			Help:      "Number of tasks in uninterruptible state",
+		},
+			[]string{"name", "id", "image"},
+		),
+
+		tasksIoWait: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "tasks_io_wait",
+			Help:      "Number of tasks waiting on IO",
+		},
+			[]string{"name", "id", "image"},
+		),
+	}
+	c.vectors = []vector{
+		c.lastSeen,
+
+		c.cpuUsageSeconds,
+		c.cpuUsageSecondsPerCPU,
+
+		c.memoryUsageBytes,
+		c.memoryWorkingSet,
+		c.memoryFailures,
+
+		c.fsLimit,
+		c.fsUsage,
+		c.fsReads,
+		c.fsReadsSectors,
+		c.fsReadsMerged,
+		c.fsReadTime,
+		c.fsWrites,
+		c.fsWritesSectors,
+		c.fsWritesMerged,
+		c.fsWriteTime,
+		c.fsIoInProgress,
+		c.fsIoTime,
+		c.fsWeightedIoTime,
+
+		c.networkRxBytes,
+		c.networkRxPackets,
+		c.networkRxErrors,
+		c.networkRxDropped,
+		c.networkTxBytes,
+		c.networkTxPackets,
+		c.networkTxErrors,
+		c.networkTxDropped,
+
+		c.tasksSleeping,
+		c.tasksRunning,
+		c.tasksStopped,
+		c.tasksUninterruptible,
+		c.tasksIoWait,
+	}
+	return c
+}
+
+// Describe describes all the metrics ever exported by cadvisor. It
+// implements prometheus.PrometheusCollector.
+func (c *PrometheusCollector) Describe(ch chan<- *prometheus.Desc) {
+	c.errors.Describe(ch)
+	for _, v := range c.vectors {
+		v.Describe(ch)
+	}
+}
+
+// Collect fetches the stats from all containers and delivers them as
+// Prometheus metrics. It implements prometheus.PrometheusCollector.
+func (c *PrometheusCollector) Collect(ch chan<- prometheus.Metric) {
+	c.mutex.Lock() // To protect metrics from concurrent collects.
+	defer c.mutex.Unlock()
+
+	for _, v := range c.vectors {
+		v.Reset()
+	}
+
+	containers, err := c.manager.SubcontainersInfo("/", &info.ContainerInfoRequest{NumStats: 1})
+	if err != nil {
+		c.errors.Inc()
```

Comment:
It is? Because this kind of error isn't about processing some items for which a error rate would be useful but something affecting overall scrape? We probably should add that to the best practices in the prometheus docs.